### PR TITLE
Refactor `KeyRange` related things

### LINKED ISSUE #46: Make `Partition` into tuple of `KeyRange` and `Block`

### LINKED ISSUE #47: Refactor `block.partition` with `groupBy`

### LINKED ISSUE #48: Define `KeyRange` class
- [ ] implement `inKeyRange`

### PR DESCRIPTION
**Pull request by betarixm**
- Closes #46
- Closes #47
- Closes #48

### CODE REVIEW

File: core/src/test/scala/BlockSpec.scala [hunk 1]
```diff
@@ -6,6 +6,17 @@ import java.io.File
 import scala.io.Source
 
 class BlockSpec extends AnyFlatSpec {
+  implicit class ComparableKeyRange(keyRange: KeyRange) {
+    def is(that: KeyRange): Boolean =
+      (keyRange.from is that.from) && (keyRange.from is that.from)
```
[the remark this answers is not in the document]
Comment (betarixm) [reply core/src/test/scala/BlockSpec.scala]: 35ff06098433be6935aec9838310c7aca8085e53

File: core/src/test/scala/BlockSpec.scala [hunk 2]
```diff
@@ -107,44 +118,44 @@ class BlockSpec extends AnyFlatSpec {
...
-    val keyStart1 = new Key(Array(0x0))
-    val keyEnd1 = new Key(Array(0x4))
-    val keyRange1 = new KeyRange(keyStart1, keyEnd1)
-    val workerMetadata1 = WorkerMetadata("1.1.1.1", 123, Option(keyRange1))
-
-    val keyStart2 = new Key(Array(0x5))
-    val keyEnd2 = new Key(Array(0x9))
-    val keyRange2 = new KeyRange(keyStart2, keyEnd2)
-    val workerMetadata2 = WorkerMetadata("2.2.2.2", 123, Option(keyRange2))
+    val firstKeyRange = KeyRange(new Key(Array(0x0)), new Key(Array(0x4)))
```
Comment (hataehyeok) [inline core/src/test/scala/BlockSpec.scala:121]: In this case, 0x4 key cannot included in any partition. I wonder that it is intended
eg.`new Record(new Key(Array(0x4)), Array(0x1, 0x2, 0x3))` never included in KeyRange

File: core/src/main/scala/KeyRange.scala [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package kr.ac.postech.paranode.core
+
+case class KeyRange(from: Key, to: Key) {
+  def includes(key: Key): Boolean = (from <= key) && (key < to)
```
Comment (hataehyeok) [inline core/src/main/scala/KeyRange.scala:4]: I wonder if it is acceptable that the final end of the KeyRange is not included in the partition. If the master sets the lowest key to the very end of the KeyRange, these records will be omitted from every partition.

I mean, below case, we cannot include 0x9 key in partitions

```scala
val firstKeyRange = KeyRange(new Key(Array(0x0)), new Key(Array(0x3)))
val secondKeyRange = KeyRange(new Key(Array(0x3)), new Key(Array(0x6)))
val thirdKeyRange = KeyRange(new Key(Array(0x6)), new Key(Array(0x9)))
```

File: core/src/test/scala/BlockSpec.scala (hunk 2 above)
Comment (betarixm) [reply core/src/test/scala/BlockSpec.scala:121]: Intended. Many range interfaces DO NOT include it. e.g. python's `ragne`, scala's `Range`.

File: core/src/main/scala/KeyRange.scala (hunk 3 above)
Comment (betarixm) [reply core/src/main/scala/KeyRange.scala:4]: Regarding `https://github.com/betarixm/434project/pull/58#discussion_r1390790192`, I want to not include the end within range. However, your comment make sense. Maybe it is okay to include end of key.

Comment (betarixm) [reply core/src/main/scala/KeyRange.scala]: 778f9fb06c45faf6efcaad42f30a44dc22df31e6

File: core/src/test/scala/BlockSpec.scala (hunk 2 above)
Comment (betarixm) [reply core/src/test/scala/BlockSpec.scala:121]: 778f9fb06c45faf6efcaad42f30a44dc22df31e6

**Review by hataehyeok — APPROVED** [review]

**Pull request merged**